packages/firestore/scripts/prune-dts.ts

```diff
@@ -0,0 +1,420 @@
+    const modifier = ts.createModifier(
+      hideConstructorTag.comment === 'protected'
+        ? ts.SyntaxKind.ProtectedKeyword
+        : ts.SyntaxKind.PrivateKeyword
+    );
+    return ts.createConstructor(
+      node.decorators,
+      [modifier],
+      /*parameters=*/ [],
+      /* body= */ undefined
+    );
+  } else {
+    return node;
+  }
+}
+
+/**
+ * Verifies that the provided private type has the same type arguments as the
+ * symbol we are trying to replace it with. Types can only be replaced directly
+ * with one another if their type arguments match.
+ */
+function verifyMatchingTypeArguments(
+  privateImportDeclaration: ts.NodeWithTypeArguments,
+  publicType: ts.ClassDeclaration | ts.InterfaceDeclaration
+): boolean {
+  const privateTypeArguments = privateImportDeclaration.typeArguments ?? [];
+  const publicTypeArguments = publicType.typeParameters ?? [];
+
+  if (privateTypeArguments.length !== publicTypeArguments.length) {
+    return false;
+  }
+
+  // Make sure that the names of the types match. We could probably be smarter
+  // about this if the types are re-ordered, but we don't need this support yet.
+  for (let i = 0; i < privateTypeArguments.length; ++i) {
+    if (
+      privateTypeArguments[i].getText() !==
+      publicTypeArguments[i].name.escapedText
+    ) {
+      return false;
+    }
+  }
+
+  return true;
+}
+
+/**
+ * Examines `extends` and `implements` clauses and removes or replaces them if
+ * they refer to a non-exported type. When an export is removed, all members
+ * from the removed class are merged into the provided class or interface
+ * declaration.
+ *
+ * @example
+ * Input:
+ * class Foo {
+ *   foo: string;
+ * }
+ * export class Bar extends Foo {}
+ *
+ * Output:
+ * export class Bar {
+ *   foo: string;
+ * }
+ */
+function prunePrivateImports<
+  T extends ts.InterfaceDeclaration | ts.ClassDeclaration
+>(
+  program: ts.Program,
+  host: ts.CompilerHost,
+  sourceFile: ts.SourceFile,
+  node: T
+): T {
+  const typeChecker = program.getTypeChecker();
+  const currentType = typeChecker.getTypeAtLocation(node);
+  const currentName = currentType?.symbol?.name;
+
+  // The list of heritage clauses after all private symbols are removed.
+  const prunedHeritageClauses: ts.HeritageClause[] = [];
+  // Additional members that are copied from the private symbols into the public
+  // symbols
+  const additionalMembers: ts.NamedDeclaration[] = [];
+
+  for (const heritageClause of node.heritageClauses || []) {
+    const exportedTypes: ts.ExpressionWithTypeArguments[] = [];
+    for (const type of heritageClause.types) {
+      if (isExported(type.modifiers)) {
+        exportedTypes.push(type);
+      } else {
+        const publicSymbol = extractPublicSymbol(
+          typeChecker,
+          sourceFile,
+          type.expression
+        );
+        if (
+          publicSymbol &&
+          publicSymbol.name !== currentName &&
+          verifyMatchingTypeArguments(
```

Comment:
I'm not sure it's the sufficient condition for a safe replacement. Considering:
```
declare class A<T> {
   aData: T;
}

export declare class B<T> extends A<string> {
    bData: T; 
}

export declare class C<T> extends A<T> {
}
```
Will it become?
```
export declare class B<T> {
    aData: string;
    bData: T; 
}
export declare class C<T> extends B<T> { // it is incorrect, because the type of `aData` becomes `string` which was `T` in the original definition.
}
```